### PR TITLE
Updating the TensorFlow models to work as expected with tokenizers v3.0.0

### PR DESCRIPTION
**Pull request by LysandreJik**
Models and tokenizers should work in harmony; this is why it is an API design choice to be able to send the output of `encode_plus` and `batch_encode_plus` straight to the model, in both PyTorch and TensorFlow:

```py
encoded_sequence = tokenizer.encode_plus(sequence)
model(encoded_sequence)  # for TensorFlow
model(**encoded_sequence)  # for PyTorch
```

With the recent changes of tokenizers-v3.0.0 and the introduction of `BatchEncoding`, the way the TensorFlow models usually identified such inputs didn't work, as it was looking for a `dict` instead of a `BatchEncoding`. This PR patches this.

This feature was previously untested; this PR addresses this by adding four different tests on each tokenizers; testing that the tokenizers return correct `BatchEncoding` objects in both PyTorch and TensorFlow which can be fed directly to the model. Both `encode_plus` and `batch_encode_plus` are tested.

Some issues were found, and were patched with this PR.

### CODE REVIEW

File: src/transformers/tokenization_transfo_xl.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ class TransfoXLTokenizer(PreTrainedTokenizer):
     vocab_files_names = VOCAB_FILES_NAMES
     pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
     max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    model_input_names = []
```
Comment (LysandreJik) [inline src/transformers/tokenization_transfo_xl.py:83]: Previously failed silently

File: tests/test_tokenization_common.py [hunk 2]
```diff
@@ -18,10 +18,31 @@
...
+        tokenizer = tokenizer_mapping[configuration][0]
+        tokenizer_fast = tokenizer_mapping[configuration][1]
+
+        model_tokenizer_mapping.update({tokenizer: (configuration, model)})
+        if tokenizer_fast is not None:
+            model_tokenizer_mapping.update({tokenizer_fast: (configuration, model)})
+
+    return model_tokenizer_mapping
+
+
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:45]: Here the arguments are between quotes as they can't be imported when a framework is not available; i.e. `TFPreTrainedModel` may not be imported unless TensorFlow >= 2.0 is installed in the environment.

File: tests/test_tokenization_common.py [hunk 3]
```diff
@@ -712,3 +733,147 @@ def _check_no_pad_token_padding(self, tokenizer, sequences):
...
+        sequence = " ".join(first_ten_tokens)
+        encoded_sequence = tokenizer.encode_plus(sequence, return_tensors="pt")
+        # This should not fail
+        model(**encoded_sequence)
+
+        if self.test_rust_tokenizer:
+            fast_tokenizer = self.get_rust_tokenizer()
+            encoded_sequence_fast = fast_tokenizer.encode_plus(sequence, return_tensors="pt")
+            # This should not fail
+            model(**encoded_sequence_fast)
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:771]: There is a lot of duplication between each test, an abstraction can certainly be done without losing too much readability. Let me know if you believe it would be better to factorize some of this code.

File: tests/test_tokenization_common.py [hunk 4]
```diff
@@ -712,3 +733,147 @@ def _check_no_pad_token_padding(self, tokenizer, sequences):
...
+
+        MODEL_TOKENIZER_MAPPING = merge_model_tokenizer_mappings(MODEL_MAPPING, TOKENIZER_MAPPING)
+
+        tokenizer = self.get_tokenizer()
+
+        if tokenizer.__class__ not in MODEL_TOKENIZER_MAPPING:
+            return
+
+        config_class, model_class = MODEL_TOKENIZER_MAPPING[tokenizer.__class__]
+        config = config_class()
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:749]: At first I wanted to initialize a configuration with only the length of the tokenizer to save memory, but doing so failed for some models which required additional configuration tweaking (like TransfoXL which required the cutoffs to be specified). This approach is the easiest in my opinion.

File: tests/test_tokenization_common.py [hunk 5]
```diff
@@ -712,3 +733,147 @@ def _check_no_pad_token_padding(self, tokenizer, sequences):
...
+
+        if config.is_encoder_decoder or config.pad_token_id is None:
+            return
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:752]: It is necessary to return when the `pad_token_id` is `None` even when not creating batches, as the `FastTokenizer` will always call `batch_encode_plus` underneath, which doesn't care about the number of sequences in the batch (even if it is 1 as called by `encode_plus`, if the tokenizer doesn't have a padding ID it will raise a `ValueError`) cc @mfuntowicz

File: tests/test_tokenization_common.py [hunk 6]
```diff
@@ -712,3 +733,147 @@ def _check_no_pad_token_padding(self, tokenizer, sequences):
...
+
+        tokenizer = self.get_tokenizer()
+
+        if tokenizer.__class__ not in MODEL_TOKENIZER_MAPPING:
+            return
+
+        config_class, model_class = MODEL_TOKENIZER_MAPPING[tokenizer.__class__]
+        config = config_class()
+
+        if config.is_encoder_decoder or config.pad_token_id is None:
```
Comment (sshleifer) [inline tests/test_tokenization_common.py:751]: Did it break for seq2seq? Happy to help

File: src/transformers/tokenization_transfo_xl.py (hunk 1 above)
Comment (thomwolf) [reply src/transformers/tokenization_transfo_xl.py:83]: 😱

File: tests/test_tokenization_common.py (hunk 4 above)
Comment (thomwolf) [reply tests/test_tokenization_common.py:749]: Good for me!

File: tests/test_tokenization_common.py [hunk 7]
```diff
@@ -712,3 +733,147 @@ def _check_no_pad_token_padding(self, tokenizer, sequences):
...
+
+        # Build sequence
+        first_ten_tokens = list(tokenizer.get_vocab().keys())[:10]
+        sequence = " ".join(first_ten_tokens)
+        encoded_sequence = tokenizer.encode_plus(sequence, return_tensors="pt")
+        # This should not fail
+        model(**encoded_sequence)
+
+        if self.test_rust_tokenizer:
+            fast_tokenizer = self.get_rust_tokenizer()
```
Comment (thomwolf) [inline tests/test_tokenization_common.py:768]: Maybe call these `get_fast_tokenizer` and `test_fast_tokenizer`?

File: tests/test_tokenization_common.py (hunk 3 above)
Comment (thomwolf) [reply tests/test_tokenization_common.py:771]: Fine for now I think

**Review by thomwolf — APPROVED** [review]
That's great!

File: tests/test_tokenization_common.py (hunk 2 above)
Comment (julien-c) [reply tests/test_tokenization_common.py:45]: - maybe you can just put the actual class names in quotes, not the Dict/Union/etc. (that's what we do elsewhere in the code, I think)
- a `PreTrainedTokenizerFast` is a `PreTrainedTokenizer` (and I don't expect that to change in the future) so you can maybe drop the first one.

**Review by julien-c — APPROVED** [review]

File: tests/test_tokenization_common.py (hunk 6 above)
Comment (LysandreJik) [reply tests/test_tokenization_common.py:751]: I believe there was no issue with BART, but there were some with T5. Namely the `decoder_input_ids` were a necessary input for the model, which is not generated for sequence to sequence models when using `encode_plus` or `batch_encode_plus`.

I think we would have to clear the expected behavior of these methods for seq2seq models before testing them.

File: tests/test_tokenization_common.py (hunk 7 above)
Comment (LysandreJik) [reply tests/test_tokenization_common.py:768]: cc @mfuntowicz
